text/0000-guesses.md

```diff
@@ -0,0 +1,258 @@
+multispan replacements are very rare (occurring exactly
+[twice in clippy](https://github.com/Manishearth/rust-clippy/search?utf8=%E2%9C%93&q=multispan_sugg))
+and therefore the API should be figured out once/if they become more common.
+
+## Changes in diagnostic API usage
+
+### Replace `suggestions` with `guesses`
+
+All current calls to `span_suggestion` are replaced with `span_guess` if 
+their replacement is obtained by a heuristic. A new type of tests is added: "suggestion". Tests in the "suggestion" directory must be failing
+with an error (or denied lint) and contain a suggestion (not required to
+be emitted by the error or lint failing the test). Then the suggestion is
+applied, and the test must now compile and run successfully.
+
+### Change `help` messages containing code snippets to `guesses`
+
+Whenever a `span_help` is passed a snippet or hand built expression,
+it is changed into a `span_guess` or `span_guesses` if multiple helps
+are generated in some form of a loop. If `span_guesses` is used, all
+arbitrary limits on the number of displayed items is removed. This limit
+is instead enforced by the command line diagnostic backend.
+
+## Json diagnostics backend API changes
+
+The json object gets a new field `guesses` which is an array of
+`Guess` objects. Every `Guess` object contains an array of span +
+snippet pairs stating which part of the code should be replaced
+with what replacment. There is no need for all guesses to replace
+the same piece of code or even require the same number of replacements.
+
+## Command line diagnostics backend API changes
+
+Currently suggestions are inserted directly into the error structure as a "help" sub-diagnostic
+of the main error, with a special `Option` field set for the replacement text.
+This means that backends like the existing command line backend and json backend will need to
+process the main error's sub-diagnostics to figure out how to treat the suggestion and extract
+that information back out of the sub-diagnostic.
+
+The backend API is changed to add a `code_hints` field to the main diagnostic, which contains
+an enum with a `Suggestion` and a `Guesses` variant. The actual backend will then destructure
+these variants and apply them as they see fit best. An implementation of this backend change
+exists [in a rust PR](https://github.com/rust-lang/rust/pull/39973) (does not implement guesses yet).
+
+The command line backend will print guesses as a sequence of "help" messages, just like in the following example
+
+```
+error[E0412]: type name `Iter` is undefined or not in scope
+ --> <anon>:4:12
+  |
+4 |     let t: Iter;
+  |            ^^^^ undefined or not in scope
+  |
+  = help: you can import several candidates into scope (`use ...;`):
+  = help:   `std::collections::binary_heap::Iter`
+  = help:   `std::collections::btree_map::Iter`
+  = help:   `std::collections::btree_set::Iter`
+  = help:   `std::collections::hash_map::Iter`
+  = help:   and 8 other candidates
```

Comment:
What's up with all these lines starting with "help:" anyway? Probably not part of this RFC but this does not look helpful to me. Starting the first line with a (highlighted) "help" should suffice.